src/core/Akka.Streams/Implementation/ChannelSinks.cs

```diff
@@ -0,0 +1,170 @@
+﻿// //-----------------------------------------------------------------------
+// // <copyright file="ChannelReaderSink.cs" company="Akka.NET Project">
+// //     Copyright (C) 2009-2022 Lightbend Inc. <http://www.lightbend.com>
+// //     Copyright (C) 2013-2022 .NET Foundation <https://github.com/akkadotnet/akka.net>
+// // </copyright>
+// //-----------------------------------------------------------------------
+
+using System;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Akka.Streams.Stage;
+
+namespace Akka.Streams.Implementation
+{
+    internal sealed class ChannelSinkLogic<T> : InGraphStageLogic
+    {
+        private readonly Inlet<T> _inlet;
+        private readonly ChannelWriter<T> _writer;
+        private readonly Action<bool> _onWriteAvailable;
+        private readonly Action<Exception> _onWriteFailed;
+        private readonly Action<Task<bool>> _onWriteReady;
+        private T _awaitingElement;
+        private readonly bool _isOwner;
+
+        public ChannelSinkLogic(SinkShape<T> shape, Inlet<T> inlet,
+            ChannelWriter<T> writer, bool isOwner) : base(shape)
+        {
+            _inlet = inlet;
+            _writer = writer;
+            _isOwner = isOwner;
+            _onWriteAvailable = GetAsyncCallback<bool>(OnWriteAvailable);
+            _onWriteFailed = GetAsyncCallback<Exception>(OnWriteFailed);
+            _onWriteReady = OnWriteReady;
+            SetHandler(_inlet, this);
+        }
+
+        private void OnWriteFailed(Exception cause) => TryFail(cause);
+
+        private void OnWriteAvailable(bool available)
+        {
+            if (available && _writer.TryWrite(_awaitingElement))
+                Pull(_inlet);
+            else
+                TryComplete();
+        }
+
+        private void TryComplete()
+        {
+            CompleteStage();
+            if (_isOwner)
+                _writer.TryComplete();
+        }
+
+        private void TryFail(Exception e)
+        {
+            FailStage(e);
+            if (_isOwner)
+                _writer.TryComplete(e);
+        }
+
+        public override void OnUpstreamFinish()
+        {
+            base.OnUpstreamFinish();
+            if (_isOwner)
+                _writer.TryComplete();
+        }
+
+        public override void OnUpstreamFailure(Exception e)
+        {
+            base.OnUpstreamFailure(e);
+            if (_isOwner)
+                _writer.TryComplete(e);
+        }
+
+        public override void PreStart()
+        {
+            Pull(_inlet);
```

Comment:
Don't think this is necessary

Reply:
The Pull on starting the sink was added as part of https://github.com/akkadotnet/Alpakka/pull/1068/ . Maybe there's a different way to fix that but without it you can't really use a sink properly.

Reply:
Ok, no worries - we'll keep it. Can always update it later.